### ISSUE TITLE
Document how to serialize using Display, deserialize using FromStr

### ISSUE THREAD

**Issue by dtolnay** [serde-rs/serde#1316]
Often a type implements Display and FromStr but not Serialize and Deserialize.

The module from https://github.com/serde-rs/json/issues/329#issuecomment-305608405 should work.

```rust
#[macro_use]
extern crate serde_derive;

extern crate serde;
extern crate serde_json;

use std::fmt::{self, Display};
use std::str::FromStr;

#[derive(Debug)]
struct X;

impl Display for X {
    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
        formatter.write_str("X")
    }
}

impl FromStr for X {
    type Err = &'static str;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "X" => Ok(X),
            _ => Err("was not X"),
        }
    }
}

#[derive(Serialize, Deserialize, Debug)]
struct S {
    #[serde(with = "string")]
    x: X
}

mod string {
    use std::fmt::Display;
    use std::str::FromStr;

    use serde::{de, Serializer, Deserialize, Deserializer};

    pub fn serialize<T, S>(value: &T, serializer: S) -> Result<S::Ok, S::Error>
    where
        T: Display,
        S: Serializer
    {
        serializer.collect_str(value)
    }

    pub fn deserialize<'de, T, D>(deserializer: D) -> Result<T, D::Error>
    where
        T: FromStr,
        T::Err: Display,
        D: Deserializer<'de>
    {
        String::deserialize(deserializer)?.parse().map_err(de::Error::custom)
    }
}

fn main() {
    let j = r#" { "x": "X" } "#;

    let s: S = serde_json::from_str(j).unwrap();
    println!("{:#?}", s);

    println!("{}", serde_json::to_string_pretty(&s).unwrap());
}
```

**Comment by stevenroose** [comment]
I came here to suggest a simple way to do exactly this. I think it would be great if serde would provide a way to do that with some kind of derive-like thing.

However, in the case of deserialization, we should try avoid the unnecessary allocation of `String`. Since `FromStr` takes a `&str`, it should be possible to use a str directly using a visitor implementation.

**Comment by koivunej** [comment]
An alternative I cooked up from the original description: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2b642a9a0b8747fe59e9190cc273810c or with some fixed typos and added deref and derefmut in [rs-ipfs/rust-ipfs/.../http/src/v0/support/serdesupport.rs](https://github.com/rs-ipfs/rust-ipfs/blob/f75af14d3f61fea3bde06202592cc2c2a29a2232/http/src/v0/support/serdesupport.rs) which might be of interest to someone. Apologies if this wrapper is already in serde, but let me know please? :)

**Comment by stevenroose** [comment]
Wouldn't it be possible to do something like this:
```rust

#[derive(Serialize, Deserialize)]
#[serde(stringify)]
pub MyType { ... }

impl fmt::Display for MyType { ... }

impl str::FromStr for MyType { ... }
```

**Comment by synek317** [comment]
> Wouldn't it be possible to do something like this:

It would be definitely useful, as far as I know, there is no concise way to use Display and FromStr for enum

**Comment by MaxVerevkin** [comment]
> `#[serde(stringify)]`

That's a really cool idea. Would be even better if this was separate for serialization and deserialization.

> there is no concise way to use Display and FromStr for enum

Can you elaborate please?

**Comment by synek317** [comment]
> Can you elaborate please?

Well, there's not much to elaborate on. I only meant that we currently need a lot of code to de/serialize using FromStr and  Display. IMHO that's a common need and it would be nice to have an attribute that would do the job for us.

**Comment by cyqsimon** [comment]
Bump. I am in favour of creating such an attribute to automate this. Should we perhaps create a new issue? Or can this issue be edited to become an "improvement suggestion" of sorts?

**Comment by silverlyra** [comment]
As noted in https://github.com/serde-rs/serde/pull/2017#issuecomment-903260861, the [`serde-with`](https://lib.rs/crates/serde_with) crate implements this as [`SerializeDisplay`](https://docs.rs/serde_with/latest/serde_with/derive.SerializeDisplay.html) and [`DeserializeFromStr`](https://docs.rs/serde_with/latest/serde_with/derive.DeserializeFromStr.html).

Using that crate, the example in the original issue:

```rs
use std::{fmt::{self, Display}, str::FromStr};

#[derive(Debug)]
struct X;

impl Display for X {
    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
        formatter.write_str("X")
    }
}

impl FromStr for X {
    type Err = &'static str;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "X" => Ok(X),
            _ => Err("was not X"),
        }
    }
}
```

can implement `Serialize` and `Deserialize` by simply adding:

```rs
use serde_with::{DeserializeFromStr, SerializeDisplay};

#[derive(DeserializeFromStr, SerializeDisplay, Debug)]
struct X;
```

**Comment by matklad** [comment]
FWIW, I've been hitting this pretty regularly across many projects. And the nature of the problem is such that, in any given project, you probably only have a couple of types, so it _doesn't_ really make sense to pull the entire `serde_with`. So, while this _can_ be implemented by a third party, using that solution is annoying, and that, for me, points towards inclusion into the core serde.

**Comment by emilyyyylime** [comment]
I would definitely be in favour of a simple `#[serde(stringify_deserialize)]` over `#[serde_as]` together with `#[serde_as(as = "DisplayFromStr")]`

**Comment by Oliboy50** [comment]
Hitting this issue again (~3rd time in 1 year) while adding a `#[derive(Serialize)]` in my `enum` would print me the name of the variant instead of using its `Display` implementation...

It would be really great to have something like a `#[derive(SerializeUsingDisplay)]` kind of  derive macro for such a simple need 🙏 🙏 🙏 

BTW, we should also answer properly [this stackoverflow question](https://stackoverflow.com/questions/58103801/how-can-i-implement-serialize-using-an-existing-display-trait-implementation) when a builtin serde solution will exist.

**Comment by Oliboy50** [comment]
ℹ️ for my "backed string enum" use case, I actually found a quiet viable alternative which is using `try_from` and `into` serde's container attributes:

```rust
#[derive(Debug, Clone, Copy, serde::Serialize, serde::Deserialize]
#[serde(try_from = "String", into = "String")] // <- add this
enum MyEnum {
    VariantA
    VariantB
}

impl Display for MyEnum {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let str = match self {
            MyEnum::VariantA => "variant_a",
            MyEnum::VariantB => "variant_b",
        };

        write!(f, "{}", str)
    }
}

impl FromStr for MyEnum {
    type Error = ();

    fn from_str(value: &str) -> Result<Self, Self::Error> {
        match value {
            "variant_a" => MyEnum::VariantA,
            "variant_b" => MyEnum::VariantB,
            _ => Err(()),
        }
    }
}

// and this
impl TryFrom<String> for MyEnum
where
    Self: FromStr,
{
    type Error = <Self as FromStr>::Err;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        MyEnum::from_str(value.as_str())
    }
}

// and this
impl From<MyEnum> for String {
    fn from(value: MyEnum) -> Self {
        value.to_string()
    }
}
```

but I don't know if this is the better way of doing it, so yes the best solution should be documented anyway

**Comment by zzhengzhuo** [comment]
Would it be better to use [SerializeDisplay](https://docs.rs/serde_with/latest/serde_with/derive.SerializeDisplay.html) and [DeserializeFromStr](https://docs.rs/serde_with/latest/serde_with/derive.DeserializeFromStr.html)?

**Comment by Oliboy50** [comment]
@zzhengzhuo yes, except [if you're not OK with adding an extra dependency for that](https://github.com/serde-rs/serde/issues/1316#issuecomment-1868540124)
